service/controller/v24/resource/ipam/create.go

```diff
@@ -212,6 +218,31 @@ func (r *Resource) selectRandomAZs(n int) ([]string, error) {
 	return shuffledAZs, nil
 }
 
+// getAvailabilityZones gets availability zones for region configured in
+// controllercontext client.
+func getAvailabilityZones(ctx context.Context) ([]string, error) {
```

Comment:
This might need to be changed so that master information comes from `installations` because it seems like there's still no way to query availability of specific instance type on given AZ in AWS so we need to manually maintain this information.

Reply:
That does not seem to be related to the region change. How is this related?

Reply:
Changes for this were required because earlier the `r.availabilityZones` were only for the same region that the control plane was running at. In order to be able to pick up correct AZs for tenant cluster in different region, we must have information of all availability zones on all regions or alternatively query them from API as is done here.

Initially I also tried adding support for removing AZs that don't provide specified instance type, but I failed at this because even today AWS doesn't provide an API that one may query for this. I tried to work around this by checking SpotInstancePriceHistory for given AZ and instance type for the last minute, but AWS didn't return any error on this even though I specified very old (`m2.large`) instance type. Hence I removed that and just left this in order to proceed with other work and get this going.

Right now we have tons of static information about AWS instances and availability zones and also little bit stuff from regions as well and I think that if we continue to work with that, it should be probably refactored to be a bit more sane so that there isn't so much repetition. This is some work though and hence I didn't start on this during hackathon.

Reply:
So effectively this fetches the AZs using the client configured with the desired region? If so, this makes sense. Do we have an overview of all the things we still have to make more dynamic?

Reply:
> So effectively this fetches the AZs using the client configured with the desired region?

Yes.

> Do we have an overview of all the things we still have to make more dynamic?

We need to somehow find a way to determine if requested instance type is available on given availability zones. Problem is that I'm not sure if this can be efficiently done dynamically.

There is a ~615MB JSON file that contains all AWS resources everywhere and their prices. From that JSON one is able to compute this information, but I think this should be something that is computed offline and committed to `installations` pre-formatted.

Reply:
I think we can make this part of the build process and at the end we get a comprehensive list of the available instances per region and AZ or something like this.